### ISSUE TITLE
How to handle errors properly?

### ISSUE THREAD

**Issue by DJDaveMark** [brianc/node-postgres#2492]
Is this still the correct way to handle errors?

We wrote this code back in January 2017 when we were using `^4.4.3` using `pg.connect` before refactoring it to use `pool.connect` when we upgraded to use `^7.11.0` back in June 2019

```javascript
pool.connect(function (err, client, done) {
    if (err) {
        done(err); // done may be called multiple times with no side-effects
        console.error('could not connect to postgres', err);
        callback(err);
        return;
    }

    // do something
});
```

today we started getting the following error:
```
TypeError: done is not a function
```
pointing to the 3rd line of the code above.

I see from the [current documentation](https://node-postgres.com/features/pooling#checkout-use-and-return) that done is not called in case of an error:
```
pool.connect((err, client, done) => {
  if (err) throw err
  client.query('SELECT * FROM users WHERE id = $1', [1], (err, res) => {
    done()
    if (err) {
      console.log(err.stack)
    } else {
      console.log(res.rows[0])
    }
  })
})
```
But I can't find any guidelines in the current documentation on how are you supposed to handle errors (in ^7.11.0 at least).

Can anyone provide an example?

**Note**: we are using old-style callbacks

**Comment by sehrope** [comment]
You should not be getting that `done is not a function` error.

If the connection creation or acquisition encounters an error then the client arg is undefined and the done callback is a no-op function. Nothing happens if you call the no-op done function but you do not have to as there is no client to release back.

What's the exact version of node, pg, and pg-pool that you are using? And is the error a connection creation one or a timeout one?

> I see from the current documentation that done is not called in case of an error:

Those examples need to be cleaned up anyway though as it looks like it's just throwing the error as an unhandled exception.

**Comment by charmander** [comment]
Correct, you aren’t supposed to call `done` in the case that `pool.connect` calls its callback with an error. Did that work once? It definitely looks like it doesn’t in the current pg-pool code.

Fixed example:

```js
pool.connect(function (err, client, done) {
    if (err) {
        console.error('could not connect to postgres', err);
        callback(err);
        return;
    }

    // do something
});
```

Remember that you should still call `done` no matter what after that point, though, because then you have the client. When calling it because of a client error, or because of any error that leaves the client in a state where it shouldn’t be reused – or, for simplicity, because of any error at all – make sure to pass a truthy value, e.g. `done(true)` or `done(err)`.

```js
pool.connect(function (err, client, done) {
    if (err) {
        console.error('could not connect to postgres', err);
        callback(err);
        return;
    }

    client.query('SELECT * FROM users WHERE id = $1', [1], function (err, res) {
        done(err);
        if (err) {
            callback(err);
        } else {
            callback(res.rows[0]);
        }
    });
});
```

**Comment by budarin** [comment]
I'm not getting error.message in error in case connection errors (for an example: connection timeout)

```
try {
    const res = await pool.query(text, params);
    return res.rows[0];
} catch (error) {
    console.error(error.message);
}
```

`error.message` is undefined
we can get error text in stack only
